sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -92,10 +92,16 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  def isCorrelatedExists(e: Expression): Boolean = e match {
+    case Exists(_, children, _) if children.isEmpty => false
```

Comment:
can we deal with `InSubquery` here as well?

Reply:
We have already rewritten uncorrelated `InSubquery` as uncorrelated `Exists` ahead of this rule. So we won't have an uncorrelated `InSubquery` here. And only `Exists` can be handled in physical plan. Maybe we'd better change the name as `isCorrelatedSubquery` to make it more clear.

Reply:
It looks to me that having a physical plan for  uncorrelated `InSubquery` is more ideal.

Reply:
@dongjoon-hyun mentioned in https://github.com/apache/spark/pull/23783 that we can rewrite all these predicate subqueries(In/Any/Some/All) as `Exists`. Here, this PR just tries to deal with uncorrelated subquery. IMHO, this may be useful for the further transformation. If we have a rule to transform all the uncorrelated subqueries to `Exists`, we only need to match Exists then.

If we have a physical plan for uncorrelated `InSubquery`, we can not collect all the rows of the subquery and check whether the left side values are **in** the result set. So we still need a way to push the 'in-condition' into the subquery in physical plan.

Reply:
is there any SQL system doing the same conversion? like Calcite?

In general if we can convert all predicate subqueries, that will be great. But if the conversion is newly invented here, we need to pay more attention to make sure it's corrected.

Reply:
OK, I will do a survey on other systems and come back soon.